include/SampleBuffer.h

```diff
@@ -86,11 +122,10 @@ class LMMS_EXPORT SampleBuffer
 	auto size() const -> size_type { return m_data.size(); }
 	auto empty() const -> bool { return m_data.empty(); }
 
-	static auto emptyBuffer() -> std::shared_ptr<const SampleBuffer>;
-
 private:
 	std::vector<sampleFrame> m_data;
-	QString m_audioFile;
+	QString m_source; //!< absolute audio file path or base64 data
+	Source m_sourceType = Source::Unknown;
```

Comment:
I like this idea, but as I tried to implement it tonight, I concluded that the way I'm identifying samples is a little problematic and needs to be more robust.

Here's the gist of what I'm thinking now:
- An audio file name describes the data source but is not the data itself
- On the other hand, base64 is the data itself but says nothing about its source
- Storing base64 data in `Source` is costly if it's a large base64 sample, and the main goal of this PR is to improve memory usage.
- Therefore, I should distinguish between the source and the data/contents, and this should be reflected in the cache data structure. How I do this will lay the groundwork for solving the more difficult and nuanced problems with samples that we talked about on Discord.

I'm currently thinking something like:
```c++
using Key1 = SampleBuffer::Source; // comparison and hashing of sample's source
using Key2 = SampleBuffer::DataId; // comparison and hashing of sample's contents
using Value = std::weak_ptr<const SampleBuffer>;

map<Key1, map<Key2, Value>> m_cache; // tentative data structure

// Note: the maps could be unordered since the keys will implement hashing functions
```

I'm almost done working out a flow chart that describes the entire process of loading a sample from this cache, and will do the same for base64. I've also mostly worked out the details of `SampleBuffer::Source` and `SampleBuffer::DataId`.

I'll implement these ideas this weekend.

Reply:
This still feels like a bit much. `map<Key1, map<Key2, Value>>` is just flat out confusing up front, dont you think? Maybe identifiying samples can use a cryptographic hash as an identifier. This hash can then be used the same way for both audio file paths and Base64 data. That said, [QCryptographicHash exists](https://doc.qt.io/qt-5/qcryptographichash.html#hash). What do you think?

An example:

Caller: Hey, I want 'foo.wav'. Do you have an entry for it?
Cache: Not sure. Let me hash it and check in my map.
Cache: Okay, I need to look for `e31dfa153e4f50bbe0adbc72b0af4635b2d8b039` (using SHA-1 here)
Cache: Found it, will return value back to caller.

> An audio file name describes the data source but is not the data itself
> On the other hand, base64 is the data itself but says nothing about its source

For a *key* to the data, do we really need to describe the actual data somehow? The key serves as an identifier. Audio file paths work for this, but as you mentioned, they only describe some kind of pointer to the actual data we need, and the same concept cannot be easily applied for base64 data. However, if we use hashes to identify the data, we can put anything into that hash function, compare it with the thing we passed, and use that as an identifier.

> Storing base64 data in Source is costly if it's a large base64 sample, and the main goal of this PR is to improve memory usage.

Easy, we store a hash which would have a fixed length.

> Therefore, I should distinguish between the source and the data/contents, and this should be reflected in the cache data structure. How I do this will lay the groundwork for solving the more difficult and nuanced problems with samples that we talked about on Discord.

`std::unordered_map<Key, Value>` may be used, where `Key` represents some hash of an identifier to the data, while `Value` stores the buffer.

Reply:
> This still feels like a bit much. map<Key1, map<Key2, Value>> is just flat out confusing up front, dont you think?

I think `Source` not always meaning "source" is already confusing. The way it's implemented now is arguably an oversimplification of what it should be, and once you start thinking in terms of "source" vs "contents" and separate those two in the cache, a lot of what we talked about on Discord becomes conceptually clearer even if it is slightly more complex in its implementation.

Data could be retrieved like this: `m_cache["foo.wav"][dataId]`

> An example

That example can't handle the case where two samples have the same file source but different contents. I'd like to at least lay the groundwork for dealing with cases like that.

Reply:
> Two samples have the same file source, but different contents.

Are you trying to say that the contents the cache returns for a certain source can change over the course of running the program (file system updates), and we need a stronger identifier for a resource?

Then lets make the identifer for contents then. If its a buffer from a path that they want, we read all the data from that resource using [`QFile::readAll`](https://doc.qt.io/qt-5/qiodevice.html#readAll) and pass that into a hash function. If its a buffer from a Base64 string that they want, we decode it using [`QByteArray::fromBase64`](https://doc.qt.io/qt-5/qbytearray.html#fromBase64) and pass that into the same hash function. Then we can check what the cache stored for that hash.

After that, I think we could remove `QFileSystemWatcher` and not have a dependency of keeping track of file system changes anymore because we are not only storing *contents* (the caching that we want to do), but also identifying *contents* uniquely (the behavior we needed). We wouldn't be storing some external handle to the content anymore that we have little control over directly and that doesn't work in all cases since you cannot assign a path to a Base64 string.

Situations for file paths after I stored `foo.wav`'s contents in a cache:
1. If `foo.wav` changes to have rickroll audio data, the hashes would change. So when a request is sent for `foo.wav` and we hash it again, no entry would be found.
2. If `foo.wav` is renamed to `bar.wav`, then it would not matter at all, so when a request is sent for `bar.wav`, we would still hash the contents in the same way and compare it with the cache.
3. If `foo.wav` is deleted, then they can't send a request for it. What I mean is that the contents of the sample they want is gone, so its just not possible. In that case, `QFile::readAll` would return an empty `QByteArray`. That would be compared in the cache, nothing would be found (empty hashes surely cant identify anything), and we return `nullptr` as usual (ideally throwing an error because failure to get something from the loader means that the cache couldn't load the requested resource, and clients *have* to respond appropriately. We shouldn't allow clients to sliently accept a `nullptr` and keep moving. This is also why I think `nullptr` isn't great here. If we throw in the constructors of `SampleBuffer`, something we have to do on failure, why not in `SampleLoader`, which uses them?)

Situations for a Base64 string:
1. They give us a Base64 string, cache decodes it using `QByteArray::fromBase64`, then hashes it and compares it with the stored entries.

The process is ultimately the same for both, its just the way that we read in data is different. Entries are still evicted when no consumers of that entry are left.

What I also like about this is that the cache will be unit testable and not require or care about the file system at all (please write tests for this if you can, we should really start getting into the habit of doing that because the next thing you know, something breaks and we have no idea why. Or, if you want, I can write them for you after you confirm that what was mentioned above should happen. I really think so.)

N.B: The job of the cache is to give clients the resources that they expect. We don't necessarily have to maintain the actual sample state was used when the file path was saved in the project. If they want a file on the file system we have to give them whats currently on the file system. If it data from a Base64 string we give them that.

> Data could be retrieved like this: m_cache["foo.wav"][dataId]

Caller: Hey, I want `foo.wav`.
Cache: Okay, sure. Whats your `dataId`?
Caller: What?

Reply:
Even better, instead of hashes, we could use something thats literally [meant for unique identification](https://doc.qt.io/qt-5/quuid.html#createUuidV5), I forgot you could make UUIDS from actual data, I've only used it to make random ones before.

Reply:
I'm scrapping the idea I had and just going with a `Source` class similar to what you suggested, but with support for hashing. For the time being, I think this is okay.

My idea of making a cache that clearly separates "source" from "data" had some issues with base64 data, since audio files would hash the `vector<sampleFrame>` data to produce a hash for the `DataId` class while base64 data would hash the base64 string instead of `vector<sampleFrame>`. I didn't like this discrepancy and it seemed like it could make it easy to misuse if I wasn't careful. I *could* fix it by decoding and hashing the base64 data every time it is loaded, though of course this would kill any performance benefit of loading base64 samples from the cache, and we would only be left with the memory usage benefits.

A better way to fix the problem would involve introducing UUIDs for base64 data in project files similar to how audio files have a file path. Not only would this help enable the robust system for sample data that I wanted, it would also likely allow faster lookup times in the cache and cut down on project file sizes.
Identical base64 samples are just copied and pasted in project files currently, so having a unique identifier for each base64 sample would allow the project files to store all base64 samples in a single section and just refer to them elsewhere using the UUID, which would save on space.

Maybe in the future I could experiment some more with these ideas, but for now your proposed `Source` class will be good enough.